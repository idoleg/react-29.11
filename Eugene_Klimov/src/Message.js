import React from "react";

const Message = ({id, name, content}) =>
    <div key={id} className="user-item">
        <span className="user-name">{name}: </span>
        <span className="user-content">{content}</span>
    </div>;

export default Message;