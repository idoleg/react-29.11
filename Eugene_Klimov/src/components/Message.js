import React from 'react';
import PropTypes from 'prop-types';

export const Message = (props) =>
  <div className="user-item">
    <span className="user-name">{props.name}: </span>
    <span className="user-content">{props.content}</span>
  </div>;

Message.propTypes = {
  name: PropTypes.string.isRequired,
  content: PropTypes.string.isRequired,
};
