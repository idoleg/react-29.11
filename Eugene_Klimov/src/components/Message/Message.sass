@import ../vars

.user-item, .bot-item
  font-size: 18px
  background-color: lightblue
  padding: $padding-left / 2
  border-radius: 10px
  margin: $padding-left / 2 10%
  display: flex
  flex-direction: column

.user-item
  align-self: flex-end

.bot-item
  align-self: flex-start
  background-color: lightcyan

.user-name
  color: gray
  font-size: 14px

.user-content
  color: #555

.user-date
  color: #aaa
  font-size: 12px
  margin-bottom: 3px
