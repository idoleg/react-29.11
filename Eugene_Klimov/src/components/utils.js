export const formatDate = () => {
  const date = new Date();
  return date.toLocaleDateString() + '  ' + date.toLocaleTimeString();
};
