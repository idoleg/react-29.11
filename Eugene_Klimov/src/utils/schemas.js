import {schema} from 'normalizr';

export const messages = new schema.Entity('messages');
export const chats = new schema.Entity('chats', {
  messageList: [messages],
});
