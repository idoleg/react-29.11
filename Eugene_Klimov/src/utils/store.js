import {applyMiddleware, createStore} from 'redux';
import {createBrowserHistory} from 'history';
import {routerMiddleware} from 'connected-react-router';
import initReducers from '../reducers';
import middlewares from '../middlewares';
import {persistStore, persistReducer} from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import autoMergeLevel2
  from 'redux-persist/lib/stateReconciler/autoMergeLevel2';

const persistConfig = {
  key: 'ReactMessenger',
  storage,
  stateReconciler: autoMergeLevel2,
  whitelist: ['chatReducer', 'messageReducer', 'profileReducer'],
};

export const history = createBrowserHistory();

function initStore() {
  const initialStore = {};

  const store = createStore(
    persistReducer(persistConfig, initReducers(history)),
    initialStore,
    applyMiddleware(routerMiddleware(history), ...middlewares),
  );

  const persistor = persistStore(store);
  return {store, persistor};
}

export default initStore();
