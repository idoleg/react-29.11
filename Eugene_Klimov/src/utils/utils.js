export const formatDate = () => {
  const date = new Date();
  return date.toLocaleDateString() + '  ' + date.toLocaleTimeString();
};

export const getNullCountObject = (obj) => {
  let count = 0;
  for (const [i] of Object.entries(obj)) {
    if (obj[i] === null) {
      count++;
    }
  }
  return count;
};
