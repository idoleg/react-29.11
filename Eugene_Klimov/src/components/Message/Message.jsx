import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {botName} from '../../utils/constants';
import DeleteOutlinedIcon from '@material-ui/icons/DeleteForeverOutlined';
import './Message.sass';
import {bindActionCreators} from 'redux';
import {deleteMessage} from '../../actions/messageActions';
import {connect} from 'react-redux';

const messageType = {
  author: PropTypes.string,
  content: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
  id: PropTypes.number,
  chatId: PropTypes.string,
  deleteMessage: PropTypes.func,
};

export class Message extends Component {
  static propTypes = messageType;

  handleDeleteMessage = () => {
    confirm('Действительно удалить сообщение?') ?
      this.props.deleteMessage(this.props.chatId, this.props.id) : null;
  };

  render() {
    const {author, content, date} = this.props;
    return (
      <div className={author === botName ? 'bot-item' : 'user-item'}>
        <div className='user-name-trash'>
          <span className='user-name'>{author || 'Anonymous'}</span>
          <DeleteOutlinedIcon className='delete-icon' viewBox='0 0 36 36'
                              onClick={this.handleDeleteMessage}
          />
        </div>
        <span className='user-date'>{date}</span>
        <span className='user-content'>{content}</span>
      </div>
    );
  }
}

const mapStateToProps = ({}) => ({});

const mapDispatchProps = (dispatch) =>
  bindActionCreators({deleteMessage}, dispatch);

export default connect(mapStateToProps, mapDispatchProps)(Message);
