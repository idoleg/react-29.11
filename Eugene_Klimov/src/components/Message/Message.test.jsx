import React from 'react';
import {Message} from './Message';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import {botName} from '../../utils/constants';

describe('<Message>', () => {
  it('renderer Message', () => {
    const props = {
      author: 'Klim',
      content: 'AnyText',
      date: '2019-12-29 13:06',
    };
    const elem = renderer.create(<Message {...props} />).toJSON;
    expect(elem()).toMatchSnapshot();
  });

  it('message with others names', () => {
    const props = {
      content: 'AnyText',
      date: '2019-12-29 13:24',
    };
    const elem = shallow(<Message {...props}/>);
    expect(elem.contains(<span
      className='user-name'>Anonymous</span>)).toBe(true);

    elem.setProps({author: botName});
    expect(elem.contains(<span
      className='user-name'>{botName}</span>)).toBe(true);
  });
});
