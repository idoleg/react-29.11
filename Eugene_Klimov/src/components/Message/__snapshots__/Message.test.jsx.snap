// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Message> renderer Message 1`] = `
<div
  className="user-item"
>
  <div
    className="user-name-trash"
  >
    <span
      className="user-name"
    >
      Klim
    </span>
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root delete-icon"
      focusable="false"
      onClick={[Function]}
      role="presentation"
      viewBox="0 0 36 36"
    >
      <path
        d="M14.12 10.47L12 12.59l-2.13-2.12-1.41 1.41L10.59 14l-2.12 2.12 1.41 1.41L12 15.41l2.12 2.12 1.41-1.41L13.41 14l2.12-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4zM6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM8 9h8v10H8V9z"
      />
    </svg>
  </div>
  <span
    className="user-date"
  >
    2019-12-29 13:06
  </span>
  <span
    className="user-content"
  >
    AnyText
  </span>
</div>
`;
